core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/ldap/UniqueIDReadOnlyLDAPUserStoreManager.java

```diff
@@ -4246,4 +4250,157 @@ private List<String> getListOfGroupNamesWithPrimaryGroup(String searchBase, Stri
+                try {
+                    answer = dirContext.search(groupSearchBase, "(objectClass=*)", searchCtls);
+                    String detectedFormat = inferLDAPTimestampFormat(answer, groupCreatedAttributeName);
+                    if (detectedFormat != null) {
+                        return detectedFormat;
+                    }
+                } finally {
+                    JNDIUtil.closeNamingEnumeration(answer);
+                }
+            }
+        } catch (NamingException e) {
+            throw new UserStoreException(
+                    ERROR_DURING_LDAP_TIMESTAMP_EXTRACTION.getMessage(),
+                    ERROR_DURING_LDAP_TIMESTAMP_EXTRACTION.getCode(), e);
+        } finally {
+            JNDIUtil.closeContext(dirContext);
+        }
+
+        // Fallback to the basic RFC 4517 Generalized Time syntax.
+        return DEFAULT_LDAP_BASIC_TIMESTAMP_FORMAT;
+    }
+
+    /**
+     * Checks whether the configured timestamp format contains a single format.
+     *
+     * This method uses the regex pattern "\\[[^\\[\\]]+\\]" to locate bracketed segments.
+     * Breakdown of the regex:
+     * - "\\[" matches a literal '['.
+     * - "[^\\[\\]]+" matches one or more characters that are NOT '[' or ']'. This ensures that any matched
+     *   segment does not contain nested brackets.
+     * - "\\]" matches a literal ']'.
+     *
+     * The method counts the number of these bracketed segments in the given string. It returns true
+     * if the count is either 0 (i.e. no optional segment, like in a plain format "uuuuMMddHHmmssX")
+     * or exactly 1, and returns false if more than one segment is detected.
+     *
+     * Examples:
+     * - "uuuuMMddHHmmssX" 
+     *      returns true (plain format, no bracketed optional part).
+     * - "uuuuMMddHHmmss[.SSS]X" 
+     *      returns true (one optional bracketed segment "[.SSS]").
+     * - "[uuuuMMddHHmmss[.SSS]X]" 
+     *      returns true. Although the outer brackets enclose the whole string, the regex only matches the inner
+     *      segment "[.SSS]", so only one valid bracketed segment is counted.
+     * - "[uuuuMMddHHmmss[.SSS][,SSS]X]" 
+     *      returns false because it detects two segments ("[.SSS]" and "[,SSS]").
+     * - "[uuuuMMddHHmmss[.SSS]X][uuuuMMddHHmmss[.SS]X]" 
+     *      returns false since it finds two separate bracketed segments (one from each outer part).
+     *
+     * @param configuredFormat The timestamp format configuration string.
+     * @return true if the configuration contains zero or one bracketed segment; false if it contains more than one.
+     */
+    private boolean hasSingleTimestampFormat(String configuredFormat) {
+
+        if (StringUtils.isBlank(configuredFormat)) {
+            return false;
+        }
+
+        // Create a Pattern object with the regex that matches a bracketed segment without nested brackets.
+        Pattern pattern = Pattern.compile("\\[[^\\[\\]]+\\]");
```

Comment:
Shall we precompile the regex pattern and keep it as a constant to use here instead of complying it for every request?

Reply:
Resolved via https://github.com/wso2/carbon-kernel/pull/4239/commits/d5d8bb16b2e30761acd99516f9460e75ab1ee9b4